qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,828 @@
+
+    Raises:
+        QiskitError: if the number of degrees of freedom of the fit is
+                     less than 1.
+
+    .. note::
+        ``sigma`` is assumed to be specified in the same units as ``ydata``
+        (absolute units). If sigma is instead specified in relative units
+        the `absolute_sigma=False` kwarg of scipy curve_fit must be used.
+        This affects the returned covariance ``pcov`` and error ``popt_err``
+        parameters via ``pcov(absolute_sigma=False) = pcov * reduced_chisq``
+        ``popt_err(absolute_sigma=False) = popt_err * sqrt(reduced_chisq)``.
+    """
+    # Check that degrees of freedom is greater than 0
+    dof = len(ydata) - len(p0)
+    if dof < 1:
+        raise QiskitError(
+            "The number of degrees of freedom of the fit data and model "
+            " (len(ydata) - len(p0)) is less than 1"
+        )
+
+    # Override scipy.curve_fit default for absolute_sigma=True
+    # if sigma is specified.
+    if sigma is not None and "absolute_sigma" not in kwargs:
+        kwargs["absolute_sigma"] = True
+
+    # Run curve fit
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(
+        f=f, xdata=xdata, ydata=ydata, sigma=sigma, p0=p0, bounds=bounds, **kwargs
+    )
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Calculate the reduced chi-squared for fit
+    yfits = f(xdata, *popt)
+    residues = (yfits - ydata) ** 2
+    if sigma is not None:
+        residues = residues / (sigma ** 2)
+    reduced_chisq = np.sum(residues) / dof
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
+
+    result = {
+        "popt": popt,
+        "popt_err": popt_err,
+        "pcov": pcov,
+        "reduced_chisq": reduced_chisq,
+        "dof": dof,
+        "xrange": xdata_range,
+    }
+
+    return AnalysisResult(result)
+
+
+def level2_probability(data: Dict[str, Any], outcome: Optional[str] = None) -> Tuple[float, float]:
+    """Return the outcome probability mean and variance.
+
+    Args:
+        data: A data dict containing count data.
+        outcome: bitstring for desired outcome probability.
+
+    Returns:
+        (p_mean, p_var) of the probability mean and variance estimated from the counts.
+
+    .. note::
+
+        This assumes a binomial distribution where :math:`K` counts
+        of the desired outcome from :math:`N` shots the
+        mean probability is :math:`p = K / N` and the variance is
+        :math:`\\sigma^2 = p (1-p) / N`.
+    """
+    # TODO fix sigma definition
+    # When the count is 100% zero (i.e. simulator), this yields sigma=0.
+    # This crashes scipy fitter when it calculates covariance matrix (zero-div error).
+
+    counts = data["counts"]
+    outcome = outcome or "1" * len(list(counts.keys())[0])
+
+    shots = sum(counts.values())
+    p_mean = counts.get(outcome, 0.0) / shots
+    p_var = p_mean * (1 - p_mean) / shots
+    return p_mean, p_var
+
+
+class FitOptions(dict):
```

Comment:
Okey. I'll check how that works. I think they are slightly different, because this is the options that is passed to the fitters (the key should conform to the fitter API, i.e. `p0`, `bounds`) but runtime options may be `tau_p0` `tau_min` `tau_max` for example. The intention of `FitOption` is to tell developer that this class is formatted by `_setup_fitting` method.